### PR TITLE
Subscription support for GraphiQL

### PR DESCRIPTION
**Pull request by junminstorage**
Recreated based on #436

See https://github.com/junminstorage/graphql-tutorial for its usage. Specifically see how to start web socket server to support subscription: https://github.com/junminstorage/graphql-tutorial/blob/master/src/index.js

Ready for review.

### CODE REVIEW

**Comment by junminstorage** [comment]
Ping @acao and @danielrearden for reviewing this PR.

**Comment by junminstorage** [comment]
I will wait for the review comments before solving the conflicts.

**Comment by nileio** [comment]
is there any update about this commit please ?

**Comment by cannc4** [comment]
Looking forward to this!

**Comment by beingminimal** [comment]
Hi, @junminstorage @acao  @danielrearden  what is the current status of this PR?

**Comment by junminstorage** [comment]
Ping @acao @danielrearden again to start the code review so we can move forward.

File: src/__tests__/http-test.ts
Comment (acao) [inline src/__tests__/http-test.ts:2273]: this it statement should be about subscriptions, correct?

**Review by acao — CHANGES_REQUESTED** [review]
looking great! can you add a section to the readme that shows a simple example implementation with subscriptions? or maybe even add an “examples” directory with your example repo? just so it’s apparent to users how to implement this server side?

we should clarify that this won’t work with lambda or other function as a service platforms.

File: src/__tests__/http-test.ts
Comment (junminstorage) [reply src/__tests__/http-test.ts]: right, I changed its statement and move this test case within the graphiql section as well.

Comment (junminstorage) [reply src/__tests__/http-test.ts]: I added the sample implementation and usage of ```subscriptionEndpoint``` in graphiql options in README.  graphiql itself doesn't work as service platform which is already explained well in the README. Hence I resolved this conversation.

**Comment by junminstorage** [comment]
@acao Please take one more look.

**Comment by junminstorage** [comment]
" or maybe even add an “examples” directory with your example repo? " 
My example repo is a full-fledged graphql tutorial-type repo. I need to simplify the sample code over there, translate its JS code to TS, which requires more time. I can do it in next PR if it is useful.

**Comment by junminstorage** [comment]
Ping @achao for review this PR.

**Comment by A-Tokyo** [comment]
@achao Can you please review this? 🚀 Thanks a bunch!

**Comment by Hongbo-Miao** [comment]
My concern is that this is using the archived library **[graphiql-subscriptions-fetcher](https://github.com/apollographql/GraphiQL-Subscriptions-Fetcher)**, not sure this is a good demo for people.

Actually, there are some better libraries to use GraphQL subscriptions such as https://github.com/enisdenjo/graphql-transport-ws

**Comment by Hongbo-Miao** [comment]
Nice job! Just add my two cents. My concern is this will add a library [graphiql-subscriptions-fetcher](https://github.com/apollographql/GraphiQL-Subscriptions-Fetcher) which is archived, not sure it is a good idea.

Also, there are some new simpler and lighter libraries to add GraphQL subscriptions such as [graphql-transport-ws](https://github.com/enisdenjo/graphql-transport-ws) which did better job than the heavy `subscriptions-transport-ws` in my experience.

**Comment by A-Tokyo** [comment]
@Hongbo-Miao Awesome find! I found graphql-transport-ws much better as well. atm I am using to implement subs and they are working like a charm. (except for a hack that needs to be done.)

@junminstorage, can we maybe use graphql-transport-ws instead? 🚀 

Example using graphql-transport-ws with express:
```js
/**
 * Adaptor between our httpServer and gql Subscription server
 *
 * This gets socket.io to work alongside gqls
 * In order to support more than one websocket on the same server, we need to wrap httpServer so that ws doesn't immediately grab the socket and then passes it to SubscriptionServer, which in turn immediately aborts the handshake if it isn't a graphql subscription.
 *
 * more info: https://github.com/apollographql/subscriptions-transport-ws/issues/751#issuecomment-674420442
 */
const _adaptServerToGqls = (
  server: Server,
  gqlsPath?: string = GRAPHQL_ENDPOINT,
): Server => {
  const gqlsServerProxy = new EventEmitter();
  server.on('listening', gqlsServerProxy.emit.bind(gqlsServerProxy));
  server.on('error', gqlsServerProxy.emit.bind(gqlsServerProxy));
  server.on('upgrade', (req: Request, ...rest: Array<any>): void => {
    /** Only upgrade the requests that belong to gql subscription server */
    if (req.url === gqlsPath) {
      gqlsServerProxy.emit('upgrade', req, ...rest);
    }
  });
  return gqlsServerProxy;
};

/**
 * Creates the Apollo Subscription Server for graphql subscriptions
 */
const createSubscriptionServer = () =>
  new SubscriptionServer(
    {
      execute,
      subscribe,
      schema,
      /** optional interval in ms to send KEEPALIVE messages to all clients */
      keepAlive: GQLS_KEEP_ALIVE ? Number(GQLS_KEEP_ALIVE) : 25000,
      onConnect: (
        connectionParams: Object,
        webSocket: WebSocket,
        context: ConnectionContext,
      ) =>
        subscriptionMiddleware(
          null,
          connectionParams,
          webSocket,
          context,
        )(() => {
          /** Reject non-authenticated/public connections */
          if (!webSocket.upgradeReq.user) {
            throw new Error('Authentication required for websocket connection');
          }
          return Promise.resolve({
            ...connectionParams,
            context: {
              rootValue: { request: webSocket.upgradeReq, response: {} },
            },
          });
        }),
      onOperation: (message, params, webSocket, context) =>
        subscriptionMiddleware(
          message,
          params,
          webSocket,
          context,
        )(() =>
          Promise.resolve({
            ...params,
            context: {
              rootValue: { request: webSocket?.upgradeReq, response: {} },
            },
          }),
        ),
      onDisconnect: () => {
        logger.debug('GQLS disconnect');
      },
    },
    {
      server: _adaptServerToGqls(httpServer, GRAPHQL_ENDPOINT),
      path: GRAPHQL_ENDPOINT,
    },
  );

/** Runs after the main http server starts listening */
const postListen = () => {
  subscriptionServer = createSubscriptionServer();
};
```

**Comment by Hongbo-Miao** [comment]
@A-Tokyo S hmm, I think you posted wrong demo. For `graphql-transport-ws`, it is just simple like

```js
import { createServer } from 'graphql-transport-ws';
import { buildSchema, execute, subscribe } from 'graphql';

const schema = buildSchema(`
  type Query {
    hello: String
  }
  type Subscription {
    greetings: String
  }
`);
    
const roots = {
  query: {
    hello: () => 'Hello World!',
  },
  subscription: {
    greetings: async function* sayHiIn5Languages() {
      for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
        yield { greetings: hi };
      }
    },
  },
};
    
createServer(
  {
    schema,
    execute,
    subscribe,
    roots,
  },
  {
    server, // Your HTTP server
    path: '/graphql',
  }
);
```

**Comment by A-Tokyo** [comment]
Hey @Hongbo-Miao

Checkout this  https://github.com/apollographql/subscriptions-transport-ws/issues/751#issuecomment-674420442 issue. 

The demo you posted won't support a secondary socket, so for example if you want both socket io and gql subs it won't work. That's why there's the extra handling in the demo I posted. To ensure the server can have parallel websockets and not be limited to just 1.

**Comment by junminstorage** [comment]
Hi All, thanks for the review. I will make the changes to use the new module for subscription. The original implementation I had  was posted here in 2018. #513

**Comment by junminstorage** [comment]
@A-Tokyo @Hongbo-Miao @acao and others, I have worked on this for last few days, the main issue I have with [graphql-transport-ws](https://github.com/enisdenjo/graphql-transport-ws) is that it doesn't come with browserfied distribution, while all of our client-side JS requires that, unless we change our build process.  

About the concerns of archived dependency is used, please note the main dependency here we used in this PR to support subscription via GraphiQL is [subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws) which is not archived. graphiql-subscriptions-fetcher is very tiny module and I won't concern much about it. Let me know if you insist I can implement and replace it here in this PR. 

For the server side, I can update the sample code section in README to use the [graphql-transport-ws](https://github.com/enisdenjo/graphql-transport-ws), which should be easy task as it is clearly documented in its README. However we lost the consistency cross client and server so developers could be confused. 

How about adding another sample code section to use it so developers can choose whatever they want?

**Comment by enisdenjo** [comment]
Hi @junminstorage and others,

author of [graphql-transport-ws](https://github.com/enisdenjo/graphql-transport-ws) here 👋 .

I made a conscious decision of not over-engineering the build process for browsers. It boils down to the fact that you will have your own, custom, distribution build process (or you already have one) anyway. Needing support for more browsers is a user facing decision - you will have to adjust the distribution for your app regardless, implicitly adjusting it for the lib too.

Keeping all this in mind, I tend to keep the lib as small and as lean as possible. I try to not over-engineer it in favour of keeping it readable, accessible and extendable. Besides, there is not much too polyfill in the lib anyway. It uses pure JS primitives that have been with us for a while now (only exception being the async iterator).

However, if you disagree, want to discuss, ask questions or simply add code - please do open an issue at [graphql-transport-ws](https://github.com/enisdenjo/graphql-transport-ws), I will do my uttermost best to help and steer decisions in favour of the community and a better DX!

Furthermore, the lib can be naturally integrated with `express-graphql` (taken straight from the lib's readme):
```ts
import https from 'https';
import express from 'express';
import { graphqlHTTP } from 'express-graphql';
import { createServer } from 'graphql-transport-ws';
import { execute, subscribe } from 'graphql';
import { schema } from 'my-graphql-schema';

// create express and middleware
const app = express();
app.use('/graphql', graphqlHTTP({ schema }));

// create a http server using express
const server = https.createServer(app);

server.listen(443, () => {
  createServer(
    {
      schema,
      execute,
      subscribe,
    },
    {
      server,
      path: '/graphql', // you can use the same path too, just use the `ws` schema
    },
  );
});
```

On the side note, I personally wouldn't integrate WebSockets in _any_ GraphQL server. The integration would end up adding more work for the maintainers, forcing the repo to keep up with the libs features and version releases. IMHO, it just feels like superfluous work given the fact of how easy integration actually is... I might be wrong (or off for `subscriptions-transport-ws`).

**Comment by A-Tokyo** [comment]
Hey @junminstorage, Thanks for your hard work!

I think there's no need to replace GraphiQL-Subscriptions-Fetcher, it's quite tiny like you mentioned.

We can update the code indeed but that would cause confusion. 

I propose proceeding in one of the following directions:
- We check with @enisdenjo if we can add better browser build support to graphql-subscriptions-ws?
- We simply go with the current state of the PR, as it's working
- We drop this PR and add documentation on how to integrate graphql-subscriptions-ws -- AND fix this issue https://github.com/apollographql/subscriptions-transport-ws/issues/751#issuecomment-674420442 in graphql-subscriptions-ws

**Comment by A-Tokyo** [comment]
Hey @enisdenjo, Thanks for the awesome work!

graphql-subscriptions-ws can indeed be integrated easily, however, developers face certain issues when already having another socket (eg: socket io). [issue in graphql-subscriptions-ws](https://github.com/apollographql/subscriptions-transport-ws/issues/751#issuecomment-674420442) - Can we find a way around this? 

I think that if the above issue is resolved, we can simply follow your proposal of not adding websocket support but simply adding to the documentation how to use graphql-subscriptions-ws. If the issue above is fixed, the demo you posted would work like a charm ⚡

**Comment by enisdenjo** [comment]
Hey @A-Tokyo,

Running on a multi WS environment is a rather simple. You can directly follow [the WS official documentation on multiple WebSockets sharing a single http/s server](https://github.com/websockets/ws#multiple-servers-sharing-a-single-https-server).

Copied directly [from my running CodeSandbox example](https://codesandbox.io/s/graphql-on-multi-websocket-server-zc1g6), it would look something like this:

```ts
const http = require("http");
const WebSocket = require("ws");
const url = require("url");
const { buildSchema, execute, subscribe } = require("graphql");
const { createServer, createClient } = require("graphql-transport-ws");

//create a server object:
const server = http.createServer(function (_, res) {
  res.write("Hello World!"); //write a response to the client

  res.end(); //end the response
});

/**
 * 2 websockets on different paths.
 * - /wave sends out waves
 * - /grapqhl serves graphql
 */
const waveWS = new WebSocket.Server({ noServer: true });
const graphqlWS = new WebSocket.Server({ noServer: true });

// wave on connect
waveWS.on("connection", function connection(ws) {
  ws.send("🌊");
});

server.on("upgrade", function upgrade(request, socket, head) {
  const pathname = url.parse(request.url).pathname;

  if (pathname === "/wave") {
    waveWS.handleUpgrade(request, socket, head, function done(ws) {
      waveWS.emit("connection", ws, request);
    });
  } else if (pathname === "/graphql") {
    graphqlWS.handleUpgrade(request, socket, head, function done(ws) {
      graphqlWS.emit("connection", ws, request);
    });
  } else {
    socket.destroy();
  }
});

// graphql over websocket
createServer(
  {
    schema: buildSchema(`
      type Query {
        hello: String
      }
    `),
    roots: {
      query: {
        hello: () => "World!"
      }
    },
    execute,
    subscribe
  },
  graphqlWS // 👈  the socket you wish to bind to
);

// start http server
server.listen(8080);
```

Furthermore, I'd like to clarify with @A-Tokyo, in order to avoid confusion, that "`graphql-subscriptions-ws`" does not exists. There are 2 GraphQL transport through WebSocket implementation out there currently:
- [`subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws) from Apollo
- [`graphql-transport-ws`](https://github.com/enisdenjo/graphql-transport-ws) from myself

**Comment by A-Tokyo** [comment]
Hey @enisdenjo, Thanks for the quick response.

Re confusion: Thanks for pointing it out! I updated the comments.

Re multiple sockets: This is quite clear here now, The issue is in [subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws) not [graphql-transport-ws](https://github.com/enisdenjo/graphql-transport-ws).

**Comment by enisdenjo** [comment]
Heads up, I've [added the multi WS example to the readme recipes](https://github.com/enisdenjo/graphql-transport-ws/commit/7c2362276b6cd15cf6a0ad2ab00cffd77a5f0e91). This might be helpful for others too.

**Comment by Urigo** [comment]
also another heads up about `subscriptions-transport-ws` and the chances of things getting fixed there, I suggest to follow the whole thread [here](https://github.com/apollographql/subscriptions-transport-ws/issues/777) and the blog post about `graphql-transport-ws` [here](https://the-guild.dev/blog/graphql-over-websockets) and come up with your own conclusions...

**Comment by sibelius** [comment]
If you are using koa instead of express

you could follow this logic https://github.com/sibelius/graphql-ws-multiserver/pull/1

and avoid `server.on('upgrade' ` custom logic

**Comment by junminstorage** [comment]
@enisdenjo @A-Tokyo thank you both for very quick response. The discussion has been very helpful, the community will appreciate your work on this. 

Keep in mind express-graphql is a middleware which doesn't mandate express server implementation. For concerns or discussions about how to use subscriptions-transport-ws or graphql-transport-ws in express, which doesn't affect the code changes here, we may delegate the discussions to either of two modules. 

Our main purpose of PR is the GraphiQL/browser-side subscription support, in terms of proposed actions mentioned [here]( https://github.com/graphql/express-graphql/pull/687#issuecomment-713435613), since this PR was opened back to 2018, the community are craving for the solution for a long time, I will suggest we move forward, then **immediately** start to work on few things: 

1. absorb all of the awesome samples above and add runnable examples to exmples/ directory here (I am happy to do that.) 
2. either changing our build process here to add browserfy step or add PR to graphql-transport-ws if @enisdenjo is ok with that.

**Comment by A-Tokyo** [comment]
@junminstorage Thanks a lot for your work!

I agree 100% :100:

Regarding subscriptions-transport-ws VS graphql-transport-ws, I think @enisdenjo's graphql-transport-ws is the best choice due to active up to date maintenance.

What's left now is to have this PR reviewed + merged THEN work on @junminstorage roadmap above ✨ 

This was quite fruitful 🍏

**Comment by junminstorage** [comment]
Ping @acao and @danielrearden , can you take a look at this PR again?

**Comment by Urigo** [comment]
another note - I actually wrote [GraphiQL-Subscriptions-Fetcher](https://github.com/apollographql/GraphiQL-Subscriptions-Fetcher) while I was still working at Apollo   :)

So if someone wants to just move it somewhere else and update it, I would be very happy to help and support you

**Comment by acao** [comment]
I will take a look very soon, it’s looking great!

**Comment by A-Tokyo** [comment]
@Urigo Would be happy to take over and maintain **IF** needed, how should we proceed?

**Comment by enisdenjo** [comment]
Just wanted to touch upon:

> either changing our build process here to add browserfy step or add PR to graphql-transport-ws if @enisdenjo is ok with that

I think that the build process should be a user facing decision and implementation. I'd honestly like to avoid mixing in the two worlds by supporting various build pipelines. Its also a never ending task, people will want more build options; but, the lib is about WebSockets, not distribution.

Discussions, opening issues and PRs are always welcome with me! I'd just like to be 100% behind the decision before I go by supporting it.

**Comment by acao** [comment]
@Urigo i think your subscriptions fetcher would make a lot of sense in the graphiql monorepo, since we were using a clone of your project anyways for examples. we use typescript compiler as the main build toolchain there, and we can add rollup. we got rid of browserify a while ago, since it doesn’t support tree shaking

**Comment by Urigo** [comment]
couldn't agree more!  
The only reason this repo existed was that back then we didn't have the amazing @acao on the project!

**Comment by enisdenjo** [comment]
Hey guys, just wanted to share that `graphql-ws` ships UMD builds too as of [v1.11.0](https://github.com/enisdenjo/graphql-ws/releases/tag/v1.11.0).

You can now rely on the lib in the browser like this:
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>GraphQL over WebSocket</title>
    <script
      type="text/javascript"
      src="https://unpkg.com/graphql-ws/umd/graphql-ws.min.js"
    ></script>
  </head>
  <body>
    <script type="text/javascript">
      (function () {
        const client = graphqlWs.createClient({
          url: 'wss://umdfor.the/win/graphql',
        });

        // consider other recipes for usage inspiration
      })();
    </script>
  </body>
</html>
```
_* Plucked straight out of [the lib recipes](https://github.com/enisdenjo/graphql-ws#recipes)_

**Comment by acao** [comment]
sounds great, I can use this for the graphiql branch CDN example, too!

**Comment by junminstorage** [comment]
@acao happy weekend! 
Will you be able to provide any direction for us to make progress on this PR? Lots of discussions in this PR are great for future and not directly address to the changes though. Let me know what I miss here.

**Review by acao — APPROVED** [review]

**Comment by acao** [comment]
`express-graphql` is free
to provide whatever fetcher it wants until `graphiql` has its own reference `buildFetcher()` library. what y’all have here looks great and i don’t see any reason to block merging. I made progress on a branch for graphiql monorepo to create this new library, but other things came up.

i also don’t think this impacts the `IncrementalDelivery` implementation in `experimental` branch, since that doesn’t touch subscriptions afaik, but worth checking into for sure

**Comment by junminstorage** [comment]
@acao thanks for approving it. I just resolved the merge conflict. Could you or someone else merge it? Let me know what is the next step.

**Comment by acao** [comment]
would it be considered a breaking change to eventually replace `subscriptions-transport-ws` with `graphql-ws`? because we were using that on the branch for the utility function

**Comment by junminstorage** [comment]
From what I read through the two repos: subscriptions-transport-ws and graphql-ws, the switch should not be breaking changes.

**Comment by enisdenjo** [comment]
Awesome to hear that you're considering `graphql-ws` as a replacement!

Just wanted to give a short disclaimer - beware that the transport Protocol is NOT interchangeable between the two libraries. If you chose one or the other, you must make sure that both the server and the client follow the designated transport protocol.

TL;DR: [`subscriptions-transport-ws` transport Protocol](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) **!=** [`graphql-ws` transport Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md)

**Comment by acao** [comment]
that's what I had thought. I'm going to ask folks working on the transport spec and the new technical review committee I think. whatever we choose here sets a massive precedent, and we want to be sure to pick the protocol that will make the most sense moving forward. backwards compatible support for servers that implement `subscriptions-transport-ws` style wss protocol is a factor as well

**Comment by enisdenjo** [comment]
Using the subprotocol(s) requested by the client in the initial WS handshake, the server can figure out which protocol is necessary and delegate the connection to the relevant lib. So, yeah, you can still support both!

PostGraphile had recently merged a PR doing exactly that: supporting both protocols by delegating connections to the appropriate destination. You can check the PR in question here: https://github.com/graphile/postgraphile/pull/1338 and the delegator here: https://github.com/graphile/postgraphile/blob/8621b19ff8a0ffd4f245466733e61606967d9c3d/src/postgraphile/http/subscriptions.ts#L460-L485

**Comment by acao** [comment]
i was just writing a message to @IvanGoncharov suggesting a strategy like this! so, i think this PR is good for now then, since most are still using the original apollo spec. we can implement this logic in a follow up PR or in this PR?

**Comment by danielrearden** [comment]
@acao @enisdenjo I whole-heartedly disagree with the notion that `express-graphql` should support both libraries. Denis has single-handedly been driving the WS RFC for the GraphQL over HTTP spec. The spec and its RFCs should be the only thing we worry about supporting and `graphql-ws` is the only library that would support that goal.

**Comment by acao** [comment]
it’s very minimal for either when it comes to express-graphql. there is no actual implementation but just some docs entries mostly.  the new protocol will be recommended for graphiql, but we should document both. @enisdenjo will provide a follow up PR for graphql-ws

**Comment by junminstorage** [comment]
Thank you all for the input. I am so glad we moved forward with the subscription support here. I also highly recommend we provided backward compatibility support, for major or big companies like mine, this is very important. 

I will continue to provide support for this feature, and @enisdenjo just let me know I can help you create follow-up PR as well.

**Comment by enisdenjo** [comment]
Hey hey guys, sorry for the wait - my time was absolutely clogged...

[`graphql-ws` now offers a backwards compatibility recipe](https://github.com/enisdenjo/graphql-ws#ws-backwards-compat) which can help ease the transitioning phase. Just 2 questions from my side to you, @junminstorage:
- Why have you [used the 4 year old version `subscriptions-transport-ws@0.5.4`](https://github.com/graphql/express-graphql/blob/6f888980b87bc45cc27d41e9150df040993d9376/package.json#L98) when the latest is [`v0.9.18`](https://www.npmjs.com/package/subscriptions-transport-ws/v/0.9.18)? _(Maybe the answer is somewhere within these comments, but I didnt bother reading them 😅)_
- Would you like to take up the task of adding `graphql-ws` with backwards compatibility? Shouldn't be a big deal following the new recipe. I just feel you'd be faster having done all this great work already. If not, I'll push a PR when my time frees up - no biggie!

**Comment by junminstorage** [comment]
@enisdenjo Thanks for the feedback. 
1. my original PR was created two years ago: #436. I forgot to upgrade.
2. I have some free time during this holiday season I am happy to work on that. 

Just to rephrase our plan here: we will switch to graphql-ws on web client side, then follow your PR at: graphile/postgraphile#1338 to add delegator on server side to have backwards compatibility support.

**Comment by junminstorage** [comment]
I made the necessary changes so we can support graphql-ws and subscriptions-transport-ws clients, however I am stuck at one issue may need @enisdenjo 's support, hence I opened an issue in his graphq-ws repo. https://github.com/enisdenjo/graphql-ws/issues/93.
Alternatively I can write another subscription fetcher.

**Comment by enisdenjo** [comment]
I've showcased a simple, minimal, `fetcher` over at the issue you opened on the lib. Check it out: https://github.com/enisdenjo/graphql-ws/issues/93#issuecomment-758620896.

Hope this helps. If you need further assistance, feel free to ask! I am more than happy to help. 😄

**Pull request merged**
